bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/graphics/Resource.java

```diff
@@ -165,29 +170,28 @@ public Device getDevice() {
 
 void ignoreNonDisposed() {
 	if (tracker != null) {
-		tracker.ignoreMe = true;
+		tracker.setReporting(false);
 	}
 }
 
 void init() {
 	if (device.tracking) device.new_Object(this);
+	if (tracker != null) tracker.setReporting(true);
 }
 
 void initNonDisposeTracking() {
 	// Color doesn't really have any resource to be leaked, ignore.
 	if (this instanceof Color) return;
 
-	// Avoid performance costs of having '.finalize()' when not tracking.
+	// Avoid performance costs of gathering the current stack trace when not tracking.
 	if (nonDisposedReporter == null) return;
 
 	// Capture a stack trace to help investigating the leak
 	Error error = new Error("SWT Resource was not properly disposed"); //$NON-NLS-1$
 
-	// Allocate a helper class with '.finalize()' in it, it will do the actual
-	// work of detecting and reporting errors. This works because Resource
-	// holds the only reference to 'ResourceTracker' and therefore the tracker
-	// is only GC'ed when Resource itself is ready to be GC'ed.
-	tracker = new ResourceTracker(this, error);
+	// Create and register the tracker to run as a cleaning action for this resource.
+	tracker = new ResourceTracker(error);
+	cleaner.register(this, tracker);
```

Comment:
`init()` looked like the more appropriate place for registering the tracker, but apparently it can be called multiple times per resource object. e.g `Image.refreshImageForZoom()`